spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/servlet/RequestSlf4jMdcAutoConfiguration.java

```diff
@@ -0,0 +1,125 @@
+import org.springframework.boot.web.servlet.FilterRegistrationBean;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.core.Ordered;
+import org.springframework.util.StringUtils;
+
+/**
+ * {@link EnableAutoConfiguration Auto-configuration} for the slf4j {@link MDC}
+ * providing the {@link HttpServletRequest} source (as 'src'), URI (as 'uri')
+ * and user name (as 'uid'). If one does not exist for these or is empty, the
+ * corresponding MDC entry will not be created. This runs as early as possible
+ * in order to make these settings available to as much logging as possible.
+ * 
+ * @author Bruce Brouwer
+ * @since 2.0
+ */
+@Configuration
+@ConditionalOnWebApplication(type = Type.SERVLET)
+@ConditionalOnClass(MDC.class)
+public class RequestSlf4jMdcAutoConfiguration {
+
+	@Bean
+	public FilterRegistrationBean<RequestSlf4jMdcFilter> requestSlf4jMdcFilter() {
+		final RequestSlf4jMdcFilter filter = new RequestSlf4jMdcFilter();
+		final FilterRegistrationBean<RequestSlf4jMdcFilter> registration = new FilterRegistrationBean<>(filter);
+		registration.setDispatcherTypes(EnumSet.of(DispatcherType.REQUEST, DispatcherType.ERROR, DispatcherType.ASYNC));
+		registration.setOrder(Ordered.HIGHEST_PRECEDENCE);
+		return registration;
+	}
+
+	public static class RequestSlf4jMdcFilter implements Filter {
+		private static final Log logger = LogFactory.getLog(RequestSlf4jMdcFilter.class);
+
+		@Override
+		public void init(final FilterConfig filterConfig) throws ServletException {
+			// nothing to configure
+		}
+
+		@Override
+		public void doFilter(final ServletRequest request, final ServletResponse response, final FilterChain chain)
+				throws IOException, ServletException {
+			try (Closeable src = putMDC("src", RequestSlf4jMdcFilter::src, request)) {
+				try (Closeable uri = putMDC("uri", RequestSlf4jMdcFilter::uri, request)) {
+					try (Closeable uid = putMDC("uid", RequestSlf4jMdcFilter::uid, request)) {
+						chain.doFilter(request, response);
+					}
+				}
+			}
+		}
+
+		private static String src(HttpServletRequest request) {
+			final String xff = request.getHeader("X-Forwarded-For");
+			return StringUtils.hasLength(xff) ? xff : request.getRemoteHost();
+		}
+
+		private static String uri(HttpServletRequest request) {
+			return request.getRequestURI();
+		}
+
+		private static String uid(HttpServletRequest request) {
+			Principal user = request.getUserPrincipal();
+			return user == null ? null : user.getName();
+		}
+
+		private static Closeable putMDC(String key, Function<HttpServletRequest, String> value,
+				ServletRequest request) {
+			if (request instanceof HttpServletRequest) {
+				try {
+					final String val = value.apply((HttpServletRequest) request);
+					if (StringUtils.hasLength(val)) {
+						return MDC.putCloseable(key, val);
+					}
+				} catch (Exception e) {
```

Comment:
This try/catch is there to make sure this filter has as little impact as possible to a request. If, for example, we can't retrieve the X-Forwarded-For header and an exception is thrown, I don't want the request to fail.